airavata-api/airavata-api-server/src/main/java/org/apache/airavata/api/server/handler/AiravataServerHandler.java

```diff
@@ -4926,9 +4965,12 @@ public boolean shareResourceWithUsers(AuthzToken authzToken, String resourceId,
                 if(userPermission.getValue().equals(ResourcePermissionType.WRITE))
                     sharingClient.shareEntityWithUsers(gatewayId, resourceId,
                             Arrays.asList(userPermission.getKey()), authzToken.getClaimsMap().get(Constants.GATEWAY_ID) + ":" + "WRITE", true);
-                else
+                else if(userPermission.getValue().equals(ResourcePermissionType.READ))
                     sharingClient.shareEntityWithUsers(gatewayId, resourceId,
                             Arrays.asList(userPermission.getKey()), authzToken.getClaimsMap().get(Constants.GATEWAY_ID) + ":" + "READ", true);
+                else
```

Comment:
Can you check for the EXEC permission here and throw an exception in the else block? And on line 5004 as well?